### PR TITLE
[SPARK-18895][TESTS] Fix resource-closing-related and path-related test failures in identified ones on Windows

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

There are several tests failing due to resource-closing-related and path-related  problems on Windows as below.

- `RPackageUtilsSuite`:

```
- build an R package from a jar end to end *** FAILED *** (1 second, 625 milliseconds)
  java.io.IOException: Unable to delete file: C:\projects\spark\target\tmp\1481729427517-0\a\dep2\d\dep2-d.jar
  at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:2279)
  at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:1653)
  at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1535)

- faulty R package shows documentation *** FAILED *** (359 milliseconds)
  java.io.IOException: Unable to delete file: C:\projects\spark\target\tmp\1481729428970-0\dep1-c.jar
  at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:2279)
  at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:1653)
  at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1535)

- SparkR zipping works properly *** FAILED *** (47 milliseconds)
  java.util.regex.PatternSyntaxException: Unknown character property name {r} near index 4

C:\projects\spark\target\tmp\1481729429282-0

    ^
  at java.util.regex.Pattern.error(Pattern.java:1955)
  at java.util.regex.Pattern.charPropertyNodeFor(Pattern.java:2781)
```

- `InputOutputMetricsSuite`:

```
- input metrics for old hadoop with coalesce *** FAILED *** (240 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input metrics with cache and coalesce *** FAILED *** (109 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input metrics for new Hadoop API with coalesce *** FAILED *** (0 milliseconds)
  java.lang.IllegalArgumentException: Wrong FS: file://C:\projects\spark\target\tmp\spark-9366ec94-dac7-4a5c-a74b-3e7594a692ab\test\InputOutputMetricsSuite.txt, expected: file:///
  at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:642)
  at org.apache.hadoop.fs.FileSystem.makeQualified(FileSystem.java:462)
  at org.apache.hadoop.fs.FilterFileSystem.makeQualified(FilterFileSystem.java:114)

- input metrics when reading text file *** FAILED *** (110 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input metrics on records read - simple *** FAILED *** (125 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input metrics on records read - more stages *** FAILED *** (110 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input metrics on records - New Hadoop API *** FAILED *** (16 milliseconds)
  java.lang.IllegalArgumentException: Wrong FS: file://C:\projects\spark\target\tmp\spark-3f10a1a4-7820-4772-b821-25fd7523bf6f\test\InputOutputMetricsSuite.txt, expected: file:///
  at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:642)
  at org.apache.hadoop.fs.FileSystem.makeQualified(FileSystem.java:462)
  at org.apache.hadoop.fs.FilterFileSystem.makeQualified(FilterFileSystem.java:114)

- input metrics on records read with cache *** FAILED *** (93 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input read/write and shuffle read/write metrics all line up *** FAILED *** (93 milliseconds)
  java.io.IOException: Not a file: file:/C:/projects/spark/core/ignored
  at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:277)
  at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
  at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)

- input metrics with interleaved reads *** FAILED *** (0 milliseconds)
  java.lang.IllegalArgumentException: Wrong FS: file://C:\projects\spark\target\tmp\spark-2638d893-e89b-47ce-acd0-bbaeee78dd9b\InputOutputMetricsSuite_cart.txt, expected: file:///
  at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:642)
  at org.apache.hadoop.fs.FileSystem.makeQualified(FileSystem.java:462)
  at org.apache.hadoop.fs.FilterFileSystem.makeQualified(FilterFileSystem.java:114)

- input metrics with old CombineFileInputFormat *** FAILED *** (157 milliseconds)
  17947 was not greater than or equal to 300000 (InputOutputMetricsSuite.scala:324)
  org.scalatest.exceptions.TestFailedException:
  at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:500)
  at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
  at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:466)

- input metrics with new CombineFileInputFormat *** FAILED *** (16 milliseconds)
  java.lang.IllegalArgumentException: Wrong FS: file://C:\projects\spark\target\tmp\spark-11920c08-19d8-4c7c-9fba-28ed72b79f80\test\InputOutputMetricsSuite.txt, expected: file:///
  at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:642)
  at org.apache.hadoop.fs.FileSystem.makeQualified(FileSystem.java:462)
  at org.apache.hadoop.fs.FilterFileSystem.makeQualified(FilterFileSystem.java:114)
```

- `ReplayListenerSuite`:

```
- End-to-end replay *** FAILED *** (121 milliseconds)
  java.io.IOException: No FileSystem for scheme: C
  at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2421)
  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2428)


- End-to-end replay with compression *** FAILED *** (516 milliseconds)
  java.io.IOException: No FileSystem for scheme: C
  at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2421)
  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2428)
  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:88)
```

- `EventLoggingListenerSuite`:

```
- End-to-end event logging *** FAILED *** (7 seconds, 435 milliseconds)
  java.io.IOException: No FileSystem for scheme: C
  at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2421)
  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2428)
  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:88)

- End-to-end event logging with compression *** FAILED *** (1 second)
  java.io.IOException: No FileSystem for scheme: C
  at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2421)
  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2428)
  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:88)

- Event log name *** FAILED *** (16 milliseconds)
  "file:/[]base-dir/app1" did not equal "file:/[C:/]base-dir/app1" (EventLoggingListenerSuite.scala:123)
  org.scalatest.exceptions.TestFailedException:
  at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:500)
  at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
  at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:466)
```

This PR proposes to fix the test failures on Windows

## How was this patch tested?

Manually tested via AppVeyor

**Before**

`RPackageUtilsSuite`: https://ci.appveyor.com/project/spark-test/spark/build/273-RPackageUtilsSuite-before
`InputOutputMetricsSuite`: https://ci.appveyor.com/project/spark-test/spark/build/272-InputOutputMetricsSuite-before
`ReplayListenerSuite`: https://ci.appveyor.com/project/spark-test/spark/build/274-ReplayListenerSuite-before
`EventLoggingListenerSuite`: https://ci.appveyor.com/project/spark-test/spark/build/275-EventLoggingListenerSuite-before

**After**

`RPackageUtilsSuite`: https://ci.appveyor.com/project/spark-test/spark/build/270-RPackageUtilsSuite
`InputOutputMetricsSuite`: https://ci.appveyor.com/project/spark-test/spark/build/271-InputOutputMetricsSuite
`ReplayListenerSuite`: https://ci.appveyor.com/project/spark-test/spark/build/277-ReplayListenerSuite-after
`EventLoggingListenerSuite`: https://ci.appveyor.com/project/spark-test/spark/build/278-EventLoggingListenerSuite-after

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala [hunk 1]
```diff
@@ -176,26 +176,31 @@ private[deploy] object RPackageUtils extends Logging {
...
-          print(s"$file contains R source code. Now installing package.", printStream, Level.INFO)
-          val rSource = extractRFolder(jar, printStream, verbose)
-          if (RUtils.rPackages.isEmpty) {
-            RUtils.rPackages = Some(Utils.createTempDir().getAbsolutePath)
-          }
-          try {
-            if (!rPackageBuilder(rSource, printStream, verbose, RUtils.rPackages.get)) {
-              print(s"ERROR: Failed to build R package in $file.", printStream)
-              print(RJarDoc, printStream)
+        Utils.tryWithSafeFinally {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: Actual change is as below:

```scala
Utils.tryWithSafeFinally {
  ...
} {
  jar.close()
}
```

File: core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala [hunk 2]
```diff
@@ -231,8 +236,10 @@ private[deploy] object RPackageUtils extends Logging {
     val zipOutputStream = new ZipOutputStream(new FileOutputStream(zipFile, false))
     try {
       filesToBundle.foreach { file =>
-        // get the relative paths for proper naming in the zip file
-        val relPath = file.getAbsolutePath.replaceFirst(dir.getAbsolutePath, "")
+        // Get the relative paths for proper naming in the zip file. Note that
+        // the separator should always be / for according to ZIP specification.
+        // `relPath` here should be, for example, "/packageTest/def.R" or "/test.R".
+        val relPath = file.toURI.toString.replaceFirst(dir.toURI.toString.stripSuffix("/"), "")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: It should always be `/` according to ZIP specification (See `4.4.17 file name: (Variable)` in https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT)

File: core/src/test/scala/org/apache/spark/scheduler/EventLoggingListenerSuite.scala [hunk 3]
```diff
@@ -119,19 +119,20 @@ class EventLoggingListenerSuite extends SparkFunSuite with LocalSparkContext wit
   }
 
   test("Event log name") {
+    val baseDirUri = Utils.resolveURI("/base-dir")
     // without compression
-    assert(s"file:/base-dir/app1" === EventLoggingListener.getLogPath(
-      Utils.resolveURI("/base-dir"), "app1", None))
+    assert(s"${baseDirUri.toString}/app1" === EventLoggingListener.getLogPath(
+      baseDirUri, "app1", None))
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/scheduler/EventLoggingListenerSuite.scala]: On Windows, it compares 

```
"file:/C:/base-dir/app1" === "file:/C:/base-dir/app1"
```

whereas on Linux and Mac,

```
"file:/base-dir/app1" === "file:/base-dir/app1"
```

**Comment by HyukjinKwon** [comment]
Build started: [TESTS] `org.apache.spark.scheduler.EventLoggingListenerSuite` [PR-16305](https://ci.appveyor.com/project/spark-test/spark/branch/4A39F6AA-CAFD-4BFD-8CB9-5BCAB193BC77)
Build started: [TESTS] `org.apache.spark.scheduler.ReplayListenerSuite` [PR-16305](https://ci.appveyor.com/project/spark-test/spark/branch/D89FD4C9-3196-4C28-A9CA-36A6F4B467AD)
Build started: [TESTS] `org.apache.spark.metrics.InputOutputMetricsSuite` [PR-16305](https://ci.appveyor.com/project/spark-test/spark/branch/E40C21AC-56C7-4DFC-9E62-2C6CFEC98172)
Build started: [TESTS] `org.apache.spark.deploy.RPackageUtilsSuite` [PR-16305](https://ci.appveyor.com/project/spark-test/spark/branch/08CE93A8-BCB1-47D9-8683-755109827A62)

Diff: https://github.com/apache/spark/compare/master...spark-test:08CE93A8-BCB1-47D9-8683-755109827A62

File: core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala (hunk 2 above)
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: cc @shivaram, could I please ask to take a look? This fixes the test, `SparkR zipping works properly` on Windows in `RPackageUtilsSuite`.

**Comment by HyukjinKwon** [comment]
cc @srowen, Could you please take a look?

**Comment by shivaram** [comment]
Thanks @HyukjinKwon I'll take a look at this tomorrow - BTW do either you or @srowen know why we see errors of the form `KeyError: -9` in Jenkins ? I also saw this on another PR [2]

```
...
  File "./dev/run-tests-jenkins.py", line 219, in main
    test_result_code, test_result_note = run_tests(tests_timeout)
  File "./dev/run-tests-jenkins.py", line 140, in run_tests
    test_result_note = ' * This patch **fails %s**.' % failure_note_by_errcode[test_result_code]
KeyError: -9
```
[2] https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70236/console for #16290

**Comment by HyukjinKwon** [comment]
https://github.com/apache/spark/pull/16307 seems too. I am not too sure but I think (IIRC) I have seen this time to time. I also want to know if these have been manually fixed by someone so far or simply it is something gone wrong in Jenkins itself.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by shivaram** [comment]
cc @shaneknapp https://github.com/apache/spark/pull/16305#issuecomment-267540707 is the comment with the Jenkins errors I was talking about

**Comment by shaneknapp** [comment]
re @shivaram -- this is totally bizarre, and nothing is popping in to mind about what could be causing this.  it's definitely not the build timeout (5 hours) or the timeout in run-tests-jenkins.py, nor are any system cleanup cron jobs running that could be responsible for this.

i'll keep poking around and trying to figure this out.

File: core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala (hunk 2 above)
Comment (shivaram) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: Yeah I think using `/` always is good. Could you write a small comment on what the `toURI` is accomplishing here (as opposed to the the `getAbsolutePath` we were using before)

**Comment by shivaram** [comment]
Had a minor question on the inline comment. Otherwise the RPackageUtils change LGTM

**Comment by shaneknapp** [comment]
test this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Ah, thank you @shivaram and @shaneknapp.

File: core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala (hunk 2 above)
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: Ah, I thought you meant writing comment in the codes.. :).it just replaces the `\` on Windows to `/`. The reason for `stripSuffix` is, it seems it has a trailing `/` when the uri is known as a directory.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: For example,

**Before**


- Windows

```scala
val a = file.getAbsolutePath // "C:\...\tmp\1481863447985-0"
val b = dir.getAbsolutePath  // "C:\...\tmp\1481863447985-0\test.R"
a.replaceFirst(b,  "")       // java.util.regex.PatternSyntaxException: Unknown character property name {r} near index 4

```
```
[info]   java.util.regex.PatternSyntaxException: Unknown character property name {r} near index 4
[info] C:\projects\spark\target\tmp\1481863447985-0
[info]     ^
[info]   at java.util.regex.Pattern.error(Pattern.java:1955)
[info]   at java.util.regex.Pattern.charPropertyNodeFor(Pattern.java:2781)
[info]   at java.util.regex.Pattern.family(Pattern.java:2736)
[info]   at java.util.regex.Pattern.sequence(Pattern.java:2076)
[info]   at java.util.regex.Pattern.expr(Pattern.java:1996)
[info]   at java.util.regex.Pattern.compile(Pattern.java:1696)
[info]   at java.util.regex.Pattern.<init>(Pattern.java:1351)
[info]   at java.util.regex.Pattern.compile(Pattern.java:1028)
[info]   at java.lang.String.replaceFirst(String.java:2178)
[info]   at org.apache.spark.deploy.RPackageUtils$$anonfun$zipRLibraries$2.apply(RPackageUtils.scala:235)
```

- Linux/Mac

```scala
val a = file.getAbsolutePath // "/var/.../T/1481938681657-0/test.R"
val b = dir.getAbsolutePath  // "/var/.../T/1481938681657-0"
a.replaceFirst(b,  "")       // "/test.R"
```


**After**


- Windows

```scala
val a = file.toURI.toString              // "file:/C:/.../tmp/1481863447985-0/test.R"
val b = dir.toURI.toString               // "file:/C:/.../tmp/1481863447985-0/"
a.replaceFirst(b.stripSuffix("/"),  "")  // "/test.R"
```

- Linux/Mac


```scala
val a = file.toURI.toString              // "file:/var/.../T/1481938681657-0/test.R"
val b = dir.toURI.toString               // "file:/var/.../T/1481938681657-0/"
a.replaceFirst(b.stripSuffix("/"),  "")  // "/test.R"
```

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala]: Sure, thanks!

**Comment by shivaram** [comment]
Thanks @HyukjinKwon - Merging into master

**Comment by HyukjinKwon** [comment]
@shaneknapp, about https://github.com/apache/spark/pull/16305#issuecomment-267665207, I know it is a too wide guess but I have seen this -9 when someone rebases multiple times before the build is started quickly and it seems making other build failed too. I have seen this three-ish times recently. This is a wild guess but I would like to note this just in case.

**Comment by shaneknapp** [comment]
yeah...  there's really not much we can do about this.  thanks for bringing
it to my attention tho.

**Pull request closed without merging**